### PR TITLE
Put sparse all reduce results to input tensors

### LINKED ISSUE #31413: Allreduce for sparse tensors is not working
```
~/Workspace$ pip show torch
Name: torch
Version: 1.3.1
```

Code for allreduce sparse tensors

```python
def run_sparse(rank, size):
    a = torch.randn(5) * (rank + 1)
    mask = a > 0  # sparse mask
    i = mask.nonzero()  # value indexes
    v = a[mask]  # sparse values
    t = torch.sparse.FloatTensor(i.t(), v, a.size())

    pprint("Before\t", t.to_dense())
    dist.all_reduce(t)
    pprint("After\t", t.to_dense())
```

Results

```
~/Workspace$ python main.py
Message from 0 Before	 tensor([0.1167, 0.0000, 0.8951, 0.4808, 0.8937])
Message from 1 Before	 tensor([0.2334, 0.0000, 1.7901, 0.9616, 1.7874])
Message from 1 After	 tensor([0.2334, 0.0000, 1.7901, 0.9616, 1.7874])
Message from 0 After	 tensor([0.1167, 0.0000, 0.8951, 0.4808, 0.8937])
```

Since the usage (sparse allreduce) is not documented yet https://github.com/pytorch/pytorch/issues/1303, I assume the usage is the same as dense tensors. However, it does not yield the expected result.

The full code snippet is attached below

```python
import os, time
import torch
import torch.distributed as dist
from torch.multiprocessing import Process

def pprint(*msg):
    print("Message from %s" % rank, *msg)

def run(rank, size):
    t = torch.ones(5) * (rank + 1)
    pprint("Before\t", t)
    dist.all_reduce(t)
    pprint("After\t", t)

def run_sparse(rank, size):
    a = torch.randn(5) * (rank + 1)
    mask = a > 0  # sparse mask
    i = mask.nonzero()  # value indexes
    v = a[mask]  # sparse values
    t = torch.sparse.FloatTensor(i.t(), v, a.size())

    pprint("Before\t", t.to_dense())
    dist.all_reduce(t)
    pprint("After\t", t.to_dense())


def init_process(rank, size, fn, backend='gloo'):
    """ Initialize the distributed environment. """
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '29500'
    dist.init_process_group(backend, rank=rank, world_size=size)
    fn(rank, size)


if __name__ == "__main__":
    size = 2
    processes = []
    for rank in range(size):
        p = Process(target=init_process, args=(rank, size, run_sparse))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32226 Put sparse all reduce results to input tensors**

right now if users call torch.dist.all_reduce() on dense tensors, outputs are put in input tensors. but if users call torch.dist.all_reduce() on sparse tensors, outputs are neither returned explicitly to users nor are put in input tensors.

To make torch.dist.all_reduce() API have same behavior on both dense tensors and sparse tensors, this diff is made to make torch.dist.all_reduce() on sparse tensors to put output in input tensors as well. This is acheived by simply calling input_sparse.copy_(output_sparse), see PR https://github.com/pytorch/pytorch/pull/32197 that implemented copy_ for sparse tensors.

close #31413

Differential Revision: [D19192952](https://our.internmc.facebook.com/intern/diff/D19192952/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19192952/)!

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: how does non_blocking work for sparse tensors? For dense CPU tensors, we will need to use pinned memory to make it non_blocking, but the in-place copy for sparse tensors could change the number of elements, which might change the storage, which might prevent us from pinning the memory??

**Comment by mrshenli** [comment]
The PR (#32197) linked in the description might be wrong, the current link points to a torch script related PR

**Comment by zhaojuanmao** [comment]
> The PR (#32197) linked in the description might be wrong, the current link points to a torch script related PR

Right, somehow my PR messed up earlier, I updated commit log with correct link, basically one sparse tensor can call copy_ to copy from another sparse tensor as long as the two tensors have the same sizes().

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (zhaojuanmao) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: I think so, for cuda work, we still need to copy back data to input tensor

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (zhaojuanmao) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: hmm, I'm not sure about pinning the memory underneath. 

But output.to(inputs[i].device(), /*non_blocking=*/true) is also kind of sparse tensor non_blocking copy?

For safety, I will change this line to "inputs[i].copy_(outputs[i], /* non_blocking */ true);"

**Comment by zhaojuanmao** [comment]
This is the PR #9005 that implemented copy_ for sparse tensors.

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (mrshenli) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: It's just me don't understand how non_blocking works for sparse tensors. If it is dense and CPU to CPU, IIUC, non_blocking is ignored. And `non_blocking` only works between CPU and CUDA.

Yep, for now, removing non_blocking should be OK.

Comment (zhaojuanmao) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: I moved this copy into synchronize() call, please see updated PR. because we have to copy btw devices only after previous non blocking copy call is finished. 

regarding how non_blocking works for sparse tensors, I searched a little bit and did not find how it works with pinned memory, but from the #9005, looks like indices, values are copied separately and they have to be coalesced first

**Comment by zhaojuanmao** [comment]
The test failures are because newly added tests are not skipped properly in non GPU and non Gloo backend test environment, will fix it

File: torch/distributed/distributed_c10d.py
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py]: irrelevant change?

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: Let's add a TODO here, this can potentially be non_blocking if we do pinned_memory correctly.

File: test/test_distributed.py
Comment (mrshenli) [inline test/test_distributed.py]: I might miss sth here, why this is only testing `tensors[0]`? What about other tensors?

File: test/common_distributed.py
Comment (mrshenli) [inline test/common_distributed.py]: Didn't look into this method, this is just moving the function and there is no changes here, right?

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks for fixing. I have some minor comments inline.

Stamped to unblock. Please fix test skips before landing. Thanks!

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (zhaojuanmao) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: it is device to device copy, I think we do not need non_blocking here?

File: test/test_distributed.py
Comment (zhaojuanmao) [reply test/test_distributed.py]: because dist_all_reduce() only accepts one tensor right now? you can see in the test I intentionally only tested for inputs with size 1.

File: test/common_distributed.py
Comment (zhaojuanmao) [reply test/common_distributed.py]: yes, just moved the function

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (mrshenli) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: I see. Yep, D2D copy will be synced on current streams on both devices.

**Pull request closed without merging**
